tools/codesigningtool/codesigningtool.py

```diff
@@ -120,16 +122,116 @@ def _certificate_fingerprint(identity):
   fingerprint = fingerprint.replace(":", "")
   return fingerprint
 
+def _certificate_common_name(cert):
+  _, subject, _ = execute.execute_and_filter_output([
+    "/usr/bin/openssl",
+    "x509",
+    "-noout",
+    "-inform",
+    "DER",
+    "-subject"
+  ], inputstr=cert, raise_on_failure=True)
+  subject = subject.strip().split('/')
+  cert_cn = [f for f in subject if "CN=" in f][0]
+  cert_cn = cert_cn.replace("CN=", "")
+  return cert_cn
 
 def _get_identities_from_provisioning_profile(mpf):
   """Iterates through all the identities in a provisioning profile, lazily."""
   for identity in mpf["DeveloperCertificates"]:
-    if not isinstance(identity, bytes):
+    cert = _certificate_data(identity)
+    yield _certificate_fingerprint(cert)
+
+def _certificate_data(cert):
+  if not isinstance(cert, bytes):
       # Old versions of plistlib return the deprecated plistlib.Data type
       # instead of bytes.
-      identity = identity.data
-    yield _certificate_fingerprint(identity)
+      cert = cert.data
+
+  return cert
+
+def _get_smartcard_tokens(xml):
+  """Get available tokens from the output of 'system_profiler SPSmartCardsDataType -xml'"""
+  tokens = [x for x in xml if x.get("_name", None) == "AVAIL_SMARTCARDS_TOKEN"]
+
+  if len(tokens) == 0:
+    return []
+
+  tokens = tokens[0].get("_items", None)
+  tokens = [x.get("_name", None) for x in tokens]
+
+  return tokens
+
+def _get_smartcard_keychain(xml):
+  """Get keychain items from the output of 'system_profiler SPSmartCardsDataType -xml'"""
+  keychain = [x for x in xml if x.get("_name", None) == "AVAIL_SMARTCARDS_KEYCHAIN"]
+
+  if len(keychain) == 0:
+    return []
+
+  keychain = keychain[0].get("_items", None)
+  return keychain
+
+def _find_smartcard_identities(identity=None):
+  """Finds smartcard identitites on the current system."""
+  ids = []
+  _, xml, _ = execute.execute_and_filter_output([
+      "/usr/sbin/system_profiler",
+      "SPSmartCardsDataType",
+      "-xml"
+  ], raise_on_failure=True)
+  xml = plistlib.loads(str.encode(xml))
+  if len(xml) == 0:
+    return []
+  xml = xml[0].get("_items", None)
```

Comment:
do we need to consider all entries in here?

Reply:
This is the top level `_items` that [contains all the info we need](https://gist.github.com/thiagohmcruz/d94b747217cd4e25deeaa5f77f47561f#file-system_profiler_output-L23). I think it's safe to ignore the rest for the purpose of this function